Fixed various errors spotted with Application Verifier

CookiePLMonster opened the pull request:
First of all, I haven't found any notes about squashing commits in the Contribution Guidelines, so if I need to do so please let me know and I'll update this PR accordingly.

This PR consists of three separate bugfixes to crashes/defects I found while running Dolphin with Application Verifier on Windows. It also contains a code sanity change I consider optional, so in case of any doubts I'll just recall the last commit.

1. `DolphinAnalytics` has been made a true singleton now, unifying its `Instance()` method with the other singletons in the codebase. Previously, it used a mutex-protected `std::shared_ptr` allegedly to ensure proper behaviour in case of concurrent access. However, since C++11 static local variables are guaranteed to be initialized in a thread safe manner, as outlined here:
https://en.cppreference.com/w/cpp/language/storage_duration#Static_local_variables
\
Additionally, not using static inline class fields here actually works around a known VS2017 compiler bug, where destructor for static inline class fields is invoked more than once - in the case of Dolphin, an already destroyed `std::shared_ptr` was destroyed again, which led to a possible crash on exit.
\
Reference to a VS2017 compiler bug, allegedly fixed in VS2019:
https://developercommunity.visualstudio.com/content/problem/300686/static-inline-class-variables-have-their-destructo.html

2. Fixed a possible infoleak/crash in `HttpRequest::Response HttpRequest::Impl::Fetch` if request fails. `ERROR_LOG` attempted to use the request body as if it was a null terminated string, but it's not - leading to a out-of-bounds read and possibly output such to a log file. I fixed it by specifying length as a format specifier, so there is no need for a temporary `std::string`.

3. Fixed a crash in `DSPTool` if `-h` or `-o` arguments were included without a following argument. As `argv[argc]` is guaranteed to be null, the code would try to assign such to an `std::string` - which is illegal.

4. ~Replaced `RtlGetVersion` with `GetVersionEx` again. Comments above this code hint this function has been removed after Win8, which is not true. I assume this change was made in the past before Dolphin was properly manifested for Windows 8.1 and Windows 10 - in which case `GetVersionEx` indeed "lied" to it. Since Dolphin is now properly manifested, there is no need for such workarounds anymore.
\
That said, if anybody thinks this change is not needed I can easily roll back this commit from the PR.~
EDIT: Following a discussion down below, this has been recalled from this PR, due to be further improved separately in the future (using a documented, non-deprecated way instead).

BhaaLseN commented:
Keeping commits separate is fine, as long as the individual commits make sense. You can also amend older commits if changes should be in there.

With that out of the way, it seems you missed a few places where Analytics are being called, particularly the NoGUI build. Our CI should help you find those places if you haven't already found them.

CookiePLMonster commented:
> With that out of the way, it seems you missed a few places where Analytics are being called, particularly the NoGUI build. Our CI should help you find those places if you haven't already found them.

Yep I noticed! On my way to fix those now.

CookiePLMonster commented:
Should be fixed now!

BhaaLseN commented:
Thanks!

Could you add some additional curlies to DSPTool? We prefer balanced braces on all statements if one of them needs them; and you just added two curly blocks in there.

Also, that `ERROR_LOG` will likely become a `std::string_view` in the future (once it switches over to `fmt`), so it might be fine using that format specifier like that - but I'll let others comment on this one.

CookiePLMonster commented:
> Could you add some additional curlies to DSPTool? We prefer balanced braces on all statements if one of them needs them; and you just added two curly blocks in there.

Done! Do note they were unbalanced previously too, as `-pm` and `-psm` had braces while the others did not - however, they were at the very end. I now added braces everywhere to make it uniform with no doubts.



> Also, that `ERROR_LOG` will likely become a `std::string_view` in the future (once it switches over to `fmt`), so it might be fine using that format specifier like that - but I'll let others comment on this one.

Sure, I added the typecast to conform to the warning (as per contribution guidelines), I guess it can be reverted once `ERROR_LOG` is modified.

Although on an unrelated note, why do responses even use vectors? This whole issue would've been void if responses were passed around as strings - after all, they are texts.

leoetlino commented on Source/Core/Core/Analytics.cpp:
```diff
@@ -237,21 +237,14 @@ void DolphinAnalytics::MakeBaseBuilder()
-    builder.AddData("win-ver-spmajor", static_cast<u32>(winver.wServicePackMajor));
-    builder.AddData("win-ver-spminor", static_cast<u32>(winver.wServicePackMinor));
-  }
+  GetVersionEx(reinterpret_cast<LPOSVERSIONINFO>(&winver));
```
I don't think we should be using GetVersionEx because it's been deprecated starting from Windows 8.1. 

>[GetVersionEx may be altered or unavailable for releases after Windows 8.1. Instead, use the Version Helper functions]

https://docs.microsoft.com/en-us/windows/desktop/api/sysinfoapi/nf-sysinfoapi-getversionexa

CookiePLMonster replied:
First a less serious answer:
https://twitter.com/Notepad_plus/status/959902540195299331

Now a more serious one:
While it is marked as such, it's definitely not going anywhere anytime soon (how would it work - removed export from the DLL? No chance), and I guess that the "deprecation" is more about the fact Version Helper functions are superior alternatives for almost all use cases of this function - bar Dolphin's one.

Dolphin is also manifested for Windows 10 properly so GetVersionEx will not lie to it, just like RtlGetVersion.

stenzek replied:
Does `RtlGetVersion` also have the returning-the-version-in-the-manifest behavior? Not that there'll be any versions other than Windows 10 soon, but that could be one reason to use it.

The MS-recommended way seems much more awkward: https://docs.microsoft.com/en-us/windows/desktop/SysInfo/getting-the-system-version

CookiePLMonster replied:
It does, yes - it was also unaffected by HighVersionLie appverifier option. So indeed, if this is considered a priority this is one reason to stick to the old code - hence my remark that I can recall it from this PR at any time.

stenzek replied:
Ah, then we probably should switch to the file-version method. But I'd be fine with making that a separate PR.

CookiePLMonster replied:
You're right, especially since this was not really found using Application Verifier, so it's kinda out of place for this PR. I'll recall this commit and revisit this again at a later point, as I agree file-version method sounds correct - and may even provide info on greater granularity that "Windows version", in an unlikely case of some Windows KB hotfix breaking things.

stenzek commented:
@dolphin-emu-bot rebuild
